keps/sig-apps/20200529-delete-priority-annotations.md

```diff
@@ -0,0 +1,291 @@
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (optional)](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This feature allows making a suggestion to the ReplicaSet controller, which pod of a Deployment should be deleted first when a scale-down event happens. This can prevent session disruption in stateful applications in a trivial manner.
+
+## Motivation
+
+For some applications, it is necessary that the application can tell Kubernetes which pod can be deleted and which replica has to be protected. The reason for this is that some applications do have stateful sessions and it is not possible to put such an application into Kubernetes because of session termination resulting from "random" down-scale. If the application is able to tell Kubernetes which of the replicas contains no/few/less important active sessions, this would solve many problems. This feature is non-disruptive to the default behaviour. Only if the annotation is existing, it will make a difference in deletion order.
+
+### Goals
+
+To recommend which pod gets deleted next of a ReplicaSet. This should help to avoid major reworks in existing applications architecture:
+* [45509](https://github.com/kubernetes/kubernetes/issues/45509) - Scale down a deployment by removing specific pods
+
+
+### Non-Goals
+
+Guaranteed (in contrast to the recommendation stated in Goals) deletion of a selected replica.
+
+## Proposal
+
+The application can set the `controller.kubernetes.io/replicaset-downscale-priority` annotation to a pod through the Kubernetes API. When a downscale event happens, the pod with the lower priority value of the previously set annotation will be deleted first. If one pod of the Deployment has no priority annotation set, it will be treated as the lowest priority.
```

Comment:
@liggitt  does it make sense to remove the "replicaset-" prefix from the API so in the future we can use it for other cases (like jobs), or in this case being explicit is better?

Reply:
This is a good point. This KEP could be generalized to support both ReplicaSet and Job controllers. Given that Job doesn't support scale / autoscale (ref [context](https://github.com/kubernetes/kubernetes/pull/60139)), I'd suggest not using `scale` in the annotation name so that we don't confuse Job users.

Reply:
Agree that anything mentioning a specific controller or an operation to perform is too constraining. I prefer to choose a naming that will describe what it is. Not what it can be used for. E.g.

- `controller.kubernetes.io/pod-priority` (I am hesitant to use term `priority` though)
- `controller.kubernetes.io/pod-ranking`
- `controller.kubernetes.io/pod-cost`
- ...

Reply:
I agree with @ingvagabund , I think we should not use priority to avoid confusion with pod priority and its semantics. `pod-cost` sounds good to me.

Reply:
Another alternative is `pod-penalty`

Reply:
`pod-cost` sounds good to me too

Reply:
pod-cost sounds good! I will change it.